### ISSUE TITLE
06/29

### ISSUE THREAD

**Issue by youngsun807** [youngsun807/WONIK_KAIST_FTC#2]
# MPC
- x = 벡터, y = 다차원 스칼라
- x* = minimizer -> loss value : solution(조합) = 1:1
- search 범위가 작으면 전체 다 계산해볼 수 있지만 현실적으로 불가
- pytorch = gradient 계산을 잘하는 툴 =
- jacobian = gradient
- learning rate scheduler = cosineannealignwarmrestarts -> 맨 처음 큰값에서 점점 작아지게 하다가 데이터셋 다시 들어오면 다시 rate 크게 높였다가 주기적으로 낮추는 방향으로 찾음
- local minima 정보 적재했다가 그것들을 마지막에 앙상블해서 사용 -> stochastic gradient ? 논문 참고
- monotonic 한 함수들끼리 increasing, decreasing 합성(연산계산)하면 기존 함수가 보존이 된다라는 Theory 존재
- 라그랑지엘, 로그 등등 최적화 solver들이 존재 -> 우리는 QP 사용(테일러 근사 사용)
- constraint가 linear 한 상태라면 QP 형태로 변형해서 사용 가능
- 특정 시점에서 QP 풀고 움직이고 다시 QP 계산하고 -> sequential QP
- 우리 프로젝트는 시간 제한이 있어 scipy 사용은 못하지만 강력한 library
- scipy - constraints = 미분 가능한 함수를 넣으면 다 사용 가능
- scipy - optional item들 정의하지 않아도 계산은 되지만 속도가 매우 느림

# mpc 함수
- gamma = 미래 측정값 현재에 영향력 제거하는?
- clamb = box contraints 밖으로 나갔을 때 다시 경계값으로 돌아오도록 함
- reduceLROnplateau = 특정 변수값이 일정 수준으로 내려갔을 때 Solve 멈추는 기능
- roll_out = RNN Roll out과 똑같다. rnn cell 처럼 사용
- tolerance = 비용이 해당 값 이하일 때 Solve 완료라고 판단
- adam optimazation 대상 = WS? -> us
- ridge = L1 regularization = 영향없는 변수 0으로 만들기
- x0 = 과거 tc 데이터, u0 = ws 과거 데이터, u = 미래 예측 step의 ws 주겟다, y = 미래 u값에 대한 tc 값
